README.md
# Theme

![badge][badge-maven] ![badge][badge-mpp] ![badge][badge-android] ![badge][badge-js] ![badge][badge-jvm]

A kotlin multiplatform theme engine

## Introduction

Ever wanted to define your theme once and port it to different platforms? This is the way

## Samples

Define your ColorPalette in multiplatform code

```kotlin
val AquaGreenPallet = ColorPalette(
    primary = Color(0xFF0E9B8A),
    primaryVariant = Color(0xFF006C5D),
    onPrimary = Color(0xFFDADADA),
    secondary = Color(0xFF12AC97),
    secondaryVariant = Color(0xFF007C69),
    onSecondary = Color(0xFF121212),
    background = Color(0xFFF5F5F5),
    onBackground = Color(0xFF121212),
    backgroundVariant = Color(0xFF006C5D),
    onBackgroundVariant = Color(0xFFDADADA),
    surface = Color(0xFFFFFFFF),
    onSurface = Color(0xFF121212),
    error = Color(0xFFA82D2D),
    onError = Color(0xFF121212),
    isLight = true
)
```

then declare/choose your typography

```kotlin
val myTypography = Typography(
    h1 = TextStyle(fontSize = 2.rem),
    h2 = TextStyle(fontSize = 1.5.rem),
    h3 = TextStyle(fontSize = 1.17.rem),
    h4 = TextStyle(fontSize = 1.0.rem),
    h5 = TextStyle(fontSize = 0.83.rem),
    h6 = TextStyle(fontSize = 0.67.rem),
    subtitle1 = TextStyle(fontSize = 1.25.rem),
    subtitle2 = TextStyle(fontSize = 1.15.rem),
    body1 = TextStyle(fontSize = 1.0.rem),
    body2 = TextStyle(fontSize = 1.0.rem),
    button = TextStyle(fontSize = 1.0.rem),
    caption = TextStyle(fontSize = 0.9.rem),
    overline = TextStyle(fontSize = 0.9.rem)
)
```

Then declare your theme

```kotlin
val AquaGreenTheme = Theme(
    name = "aQua Green",
    color = AquaGreenPallet,
    text = Typography()
)
```

## Setup:Gradle

```kotlin
dependencies {
    implementation("tz.co.asoft:theme-core:+") // please use the latest version possible
}
```

## Default Themes

The engine comes with 2 default `ColorPalette` and a default `Typography` per platform/framework/library

### AquaGreenTheme

The values are set as shown on the code above

### DarkGrayTheme

Whose color palette is as bellow

```kotlin
val DarkGrayPallet = ColorPalette(
    primary = Color(0xFF1F3D4C),
    primaryVariant = Color(0xFF455A64),
    onPrimary = Color(0xFFDADADA),
    secondary = Color(0xFF12AC97),
    secondaryVariant = Color(0xFF007C69),
    onSecondary = Color(0xFF121212),
    background = Color(0xFF080808),
    onBackground = Color(0xFFFFFFFF),
    backgroundVariant = Color(0xFF121212),
    onBackgroundVariant = Color(0xFFFFFFFF),
    surface = Color(0xFF17181A),
    onSurface = Color(0xFFFFFFFF),
    error = Color(0xFFA82D2D),
    onError = Color(0xFF121212),
    isLight = false
)
``` 

### ColorPalette Extensions

While you can craft your own themes, you'll often need to just extend a theme. You can do it simply just like

```kotlin
val PaleGreenPalette = AquaGreenPalette.copy(
    primary = Color(0xFF98FB98)
)
```

### Using with kotlin-react

A react specific artifact is available for targeting `kotlin-react` specifically

#### Sample

```kotlin
    fun RBuilder.TestComp(myTheme: ReactTheme) = ThemeProvider(myTheme) {
    div {
        +"Test sample"
    }
    ThemeConsumer { theme ->
        h1 {
            attrs["background"] = theme.color.primary()
        }
    }
}
```

#### Setup: Gradle

```kotlin
dependencies {
    implementation("tz.co.asoft:theme-react:0.0.70") // please use the latest version possible
}
```

### Comming Soon

#### Theme Compose

We are working on porting the theme lib to jetpack compose as well, that way you can have one theme on web, android and
desktop

[badge-maven]: https://img.shields.io/maven-central/v/tz.co.asoft/theme-core/0.0.70?style=flat

[badge-mpp]: https://img.shields.io/badge/kotlin-multiplatform-blue?style=flat

[badge-android]: http://img.shields.io/badge/platform-android-brightgreen.svg?style=flat

[badge-js]: http://img.shields.io/badge/platform-js-yellow.svg?style=flat

[badge-jvm]: http://img.shields.io/badge/platform-jvm-orange.svg?style=flat


CHANGELOG.md
# 0.0.70 : 2021.05.02
- Upgraded builders from 1.3.24 to 1.3.30
- Supported kotlin 1.5.0

# 0.0.60 : 2021.03.24

- Changed package name from `tz.co.asoft` to `theme`
- Added `ios` and `linux` targets
- Updated `color` from `0.0.20` to `0.0.30`

# 0.0.50 : 2021.11.25

- Updated gradle from 6.7.1 to 6.8.1
- Updated builders from 1.0.0 to 1.3.0
- Updated color from 0.0.4 to 0.0.50
- Updated kotlin-react from 17.0.0-pre.129-kotlin-1.4.20 to 17.0.1-pre.141-kotlin-1.4.21
- Updated kotlin-styled from 5.2.0-pre.129-kotlin-1.4.20 to 5.2.0-pre.141-kotlin-1.4.21

# 0.0.4 : 2020.11.27

## Bug fixes

- Fixed not supporting IR backend

# 0.0.3 : 2020.11.26

## Enhancements

- Removed Iconography as it makes the syntax cumbersome

# 0.0.2 : 2020.11.26

## Bug Fixes

- Marked ThemeProvider with @JsExport to allow proper DCE

## New Artifacts

- Added `theme-css` supporting jvm and js

## New Features

- Added Iconography to theme

## Dependencies

- updated kotlin to version 1.4.20

# 0.0.1

## Availability

- Published to maven central

## Build Src

- Updated to gradle version 6.7.1

## New Features

### theme-core-metadata

- Added `ColorPallet` class
- Added `Theme<T>` class

### theme-core-js

- Added `TextStyle` class for any web framework
- Added `Typograpgy` class for any web framework
- Added `operator fun Color.invoke()` to make it easy convert to CSS colors

### theme-react

- Added `TextStyle`
- Added `ReactTheme` type alias of `Theme<Typography>`
- Added `ThemeContext`
- Added `fun RBuilder.ThemeProvider(...)`
- Added `fun RBuilder.ThemeConsumer(...)`

## Documentation

- Update readme

## Samples

- Added samples

theme-css/src/commonMain/kotlin/theme/Typography.kt
package theme

import kotlinx.css.rem

data class Typography(
    val name: String = "Basic",
    val h1: TextStyle = TextStyle(fontSize = 2.rem),
    val h2: TextStyle = TextStyle(fontSize = 1.5.rem),
    val h3: TextStyle = TextStyle(fontSize = 1.17.rem),
    val h4: TextStyle = TextStyle(fontSize = 1.0.rem),
    val h5: TextStyle = TextStyle(fontSize = 0.83.rem),
    val h6: TextStyle = TextStyle(fontSize = 0.67.rem),
    val subtitle1: TextStyle = TextStyle(fontSize = 1.25.rem),
    val subtitle2: TextStyle = TextStyle(fontSize = 1.15.rem),
    val body1: TextStyle = TextStyle(fontSize = 1.0.rem),
    val body2: TextStyle = TextStyle(fontSize = 1.0.rem),
    val button: TextStyle = TextStyle(fontSize = 1.0.rem),
    val caption: TextStyle = TextStyle(fontSize = 0.9.rem),
    val overline: TextStyle = TextStyle(fontSize = 0.9.rem)
)